### DISCUSSION TITLE
Table creation creates transaction

### DISCUSSION THREAD

**Discussion by adaamz** [sqlalchemy/sqlalchemy#11826] [Usage Questions] [answered]
Hello,
I would like to create table if it doesn't exist, if it exists it should be truncated immediately.

I have this code
```python
import sqlalchemy as sa


meta = sa.MetaData()

my_table = sa.Table(
    "my_table", meta, sa.Column("some_time", sa.DATETIME())
)


url = "mysql+pymysql://root:password@mariadb/my_database"
engine = sa.create_engine(url)

with engine.connect() as connection:
    print(f"before table creation: {connection.in_transaction()=}")
    my_table.create(connection, checkfirst=True)
    print(f"after table creation: {connection.in_transaction()=}")

    with connection.begin():
        connection.execute(my_table.delete())
```
which kinda worked in sqlalchemy 1.4 but with migration to sqlalchemy 2.0.32 it throws this error instead:
```
# python3 reproducer.py 
before table creation: connection.in_transaction()=False
after table creation: connection.in_transaction()=True
Traceback (most recent call last):
  File "/home/afs/app/reproducer.py", line 19, in <module>
    with connection.begin():
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 864, in begin
    raise exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: This connection has already initialized a SQLAlchemy Transaction() object via begin() or autobegin; can't call begin() here unless rollback() or commit() is called first.

```

Is there a way how to do it with sqlalchemy 2.0 (and sqlalchemy 1.4 as well)?
Thanks 🙂

**Comment by zzzeek** [comment, marked as answer]
in the [Python DBAPI](https://peps.python.org/pep-0249/) anytime you run a statement on a connection a transaction is started implicitly.    2.0's API makes you more aware of this.  so you could simply remove the `begin()` in your code to have the identical transactional semantics as what you had in 1.4.    OTOH, if you want the point of your `begin()` to be where a rollback should end up, then you'd want to call `connection.commit()` right after your `create()` call.

Full description of whats happening here is in the tutorial at https://docs.sqlalchemy.org/en/20/tutorial/dbapi_transactions.html#committing-changes

**Reply by adaamz** [reply]
Thanks for the answer.
It looks a bit intuitive for me, but I will use the explicit `connection.commit()` most probably.
